New module: Bamdst

vladsavelyev opened the pull request:
Module for Bamdst - resolves https://github.com/ewels/MultiQC/issues/2158

- [x] There is example tool output for tools in the <https://github.com/ewels/MultiQC_TestData> repository or attached to this PR
- [x] Code is tested and works locally (including with `--strict` flag)
- [x] `docs/modulename.md` is created
- [x] Everything that can be represented with a plot instead of a table is a plot
- [x] Report sections have a description and help text (with `self.add_section`)
- [x] There aren't any huge tables with > 6 columns (explain reasoning if so)
- [x] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red)
- [x] Module does not do any significant computational work

Parses `coverage.report` to add to general stats and a separate table with all metrics included.

Also parse `chromosomes.report` to generate 2 plots:

<img width="858" alt="Screenshot 2023-12-13 at 15 20 27" src="https://github.com/ewels/MultiQC/assets/1575412/cf51799e-468b-4db7-941c-755e71b3fdc9">

To avoid clutter, chromosome filtering patterns `include_contigs`/`exclude_contigs` are enabled, and can be overridden with the config:

```yaml
bamdst:
  exclude_contigs:
    - "*_alt"
    - "*_decoy"
    - "*_random"
    - "*_fix"
    - "HLA*"
    - "chrUn*"
    - "chrM"
    - "chrEBV"
```

Additionally, added an option to exclude contigs by total fraction, like it's implemented in mosdepth:

```yaml
bamdst:
  # Should be a fraction, e.g. 0.001 (exclude contigs with 0.1% coverage of sum of
  # coverages across all contigs)
  perchrom_fraction_cutoff: 0.001
```

Similarly to mosdepth, if all samples come with a single contig, the module will build a bar graph:

<img width="755" alt="Screenshot 2023-12-13 at 15 28 06" src="https://github.com/ewels/MultiQC/assets/1575412/52a5f1f1-a42f-4a22-b40e-7042a9877c22">

hongtir commented:
Thanks a lot!

tamuanand commented:
Hi @vladsavelyev  and @hongtir 

Please check out this comment before merging - https://github.com/ewels/MultiQC/issues/2158#issuecomment-1801675805

vladsavelyev commented:
Added a separate table with all metrics included.

Now also parsing `chromosomes.report` to generate 2 plots:

<img width="858" alt="Screenshot 2023-12-13 at 15 20 27" src="https://github.com/ewels/MultiQC/assets/1575412/cf51799e-468b-4db7-941c-755e71b3fdc9">

To avoid clutter, chromosome filtering patterns `include_contigs`/`exclude_contigs` are enabled, and can be overridden with the config:

```yaml
bamdst:
  exclude_contigs:
    - "*_alt"
    - "*_decoy"
    - "*_random"
    - "*_fix"
    - "HLA*"
    - "chrUn*"
    - "chrM"
    - "chrEBV"
```

Additionally, added an option to exclude contigs by total fraction, like it's implemented in mosdepth:

```yaml
bamdst:
  # Should be a fraction, e.g. 0.001 (exclude contigs with 0.1% coverage of sum of
  # coverages across all contigs)
  perchrom_fraction_cutoff: 0.001
```

Similarly to mosdepth, if all samples come with a single contig, the module will build a bar graph:

<img width="755" alt="Screenshot 2023-12-13 at 15 28 06" src="https://github.com/ewels/MultiQC/assets/1575412/52a5f1f1-a42f-4a22-b40e-7042a9877c22">

ewels commented on multiqc/modules/bamdst/bamdst.py:
```diff
@@ -0,0 +1,419 @@
+                "hidden": short != ">0x",
+            }
+
+        self.general_stats_addcols(data_by_sample, headers)
```
As with the other module, I don't love just sticking all columns into General Stats. If we do this a lot, or if there are many samples, it'll bloat the report size. It also makes the "Configure Columns" dialogue massive and difficult to use if there are hundreds of columns to choose from..

ewels commented on multiqc/utils/search_patterns.yaml:
```diff
@@ -16,6 +16,9 @@ anglerfish:
   contents: "Bakta:"
+bamdst/coverage:
+  contents: "## The file was created by bamdst"
+  num_lines: 1
```
In case of unexpected crap from elsewhere
```suggestion
  num_lines: 5
```

ewels commented on multiqc/utils/config_defaults.yaml:
```diff
@@ -394,6 +394,10 @@ module_order:
       module_tag:
         - DNA
         - RNA
+  - bamdst:
+      module_tag:
+        - DNA
+        - RNA
```
We should really ditch these, they're pointless.

ewels commented on multiqc/modules/bamdst/bamdst.py:
```diff
@@ -0,0 +1,419 @@
+import os
+import fnmatch
+
+import math
```
Weird, why has Ruff moved this here, as if it's not a core Python library?

ewels commented on multiqc/modules/bamdst/bamdst.py:
```diff
@@ -0,0 +1,419 @@
+                data_by_sample[f["s_name"]] = data
+
+                # Parse "chromosomes.report" sitting next to the "coverage.report" file:
+                if os.path.exists(chroms_path := os.path.join(f["root"], "chromosomes.report")):
```
Refactoring idea for MultiQC v2 - generally switch from using `os.path` to the `Path` library. `f["root"]` could then be a `Path` object, meaning that this kind of syntax could be simplified.

Instead of:

```python
os.path.join(f["root"], "chromosomes.report")
```

We could simply have:
```python
f["root"] / "chromosomes.report"
```

ewels commented on multiqc/modules/bamdst/bamdst.py:
```diff
@@ -0,0 +1,419 @@
+            if line.startswith("## Files : "):
+                # Split a line like "## Files : example/test1.bam example/test2.bam "
+                # Cannot split by spaces because the file names can contain spaces as well
+                path_line = line.split(":")[1].strip().replace(".bam", "<EXT>").replace(".cram", "<EXT>")
```
`.sam`?

ewels replied:
Any idea what shows if the tool is used in "pipeline mode"?

From the readme:

> Pipeline mode:
>
> ```bash
> samtools view in1.bam -u | bamdst -p x.bed -o ./ -
> ```

vladsavelyev replied:
Good point! Added a test case for this. In this case, the log file name will be used for the sample name i.e. (`coverage.report` -> `coverage`), so the user will have to run the tool with `--dirs`. I updated the docs for that

ewels commented on multiqc/modules/bamdst/bamdst.py:
```diff
@@ -0,0 +1,419 @@
+                # Cannot split by spaces because the file names can contain spaces as well
+                path_line = line.split(":")[1].strip().replace(".bam", "<EXT>").replace(".cram", "<EXT>")
+                names = [self.clean_s_name(sn) for sn in path_line.split("<EXT>") if sn]
+                f["s_name"] = "-".join(names)  # concatenating file names
```
Maybe tweak the code to not overwrite `f["s_name"]` if no valid filenames are found, so that it falls back to using the filename.

```suggestion
                if names:
                    f["s_name"] = "-".join(names)  # concatenating file names
```

ewels reviewed (changes requested):
Is there a reason that the report I get only has `ST0217_Lg` present in the plots?
I can see other `coverage.report` files, so I was expecting more samples to be plotted..

Other than that, a few minor comments but mostly good to go 👍🏻

vladsavelyev commented:
> Is there a reason that the report I get only has ST0217_Lg present in the plots? I can see other coverage.report files, so I was expecting more samples to be plotted..

There are default rules to exclude alt contigs (similar to `mosdepth`) and keep only chr1 ... chrY. Other samples have only data for chrM, which is excluded. 

To make it more transparent, I now added tabs to the plots so users can switch between showing all contigs, and showing only the main ones.

ewels reviewed (approved):
Looks good, thanks!
